### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by jensens**

### CODE REVIEW

File: src/node/ext/ldap/_node.py [hunk 1]
```diff
@@ -524,8 +498,10 @@ def search(self, queryFilter=None, criteria=None, attrlist=None,
                             resattr[decode(k)] = decode(v)
                 if 'dn' in attrlist:
                     resattr[u'dn'] = decode(dn)
+                # XXX: optional RDN, DN instead of key
```
Comment (rnixx) [inline src/node/ext/ldap/_node.py]: Should this go into 1.0?

File: src/node/ext/ldap/_node.py [hunk 2]
```diff
@@ -524,8 +498,10 @@ def search(self, queryFilter=None, criteria=None, attrlist=None,
...
                 res.append((key, resattr))
             else:
+                # XXX: return DN instead of key
```
Comment (rnixx) [inline src/node/ext/ldap/_node.py]: Should this go into 1.0?

File: src/node/ext/ldap/_node.py [hunk 3]
```diff
@@ -554,7 +528,7 @@ def invalidate(self, key=None):
                                    u"changed node.")
             self.storage.clear()
             self.attrs.load()
-            self._init_keys()
+            # XXX: needs to get unset again somwhere
```
Comment (rnixx) [inline src/node/ext/ldap/_node.py]: definitely should go into 1.0, but fine for beta right now

File: src/node/ext/ldap/tests.py [hunk 4]
```diff
@@ -21,7 +21,7 @@
     ('ugm/posixGroups.rst', testing.LDIF_posixGroups),
     ('ugm/sambaUsers.rst', testing.LDIF_sambaUsers),
     ('ugm/defaults.rst', testing.LDIF_data),
-    ('../../../../README.rst', testing.LDIF_data),
+    #('../../../../README.rst', testing.LDIF_data),
```
Comment (rnixx) [inline src/node/ext/ldap/tests.py]: README doctest should pass before merge

File: src/node/ext/ldap/ugm/_api.py [hunk 5]
```diff
@@ -505,10 +466,30 @@ def __getitem__(self, key):
...
         except KeyError:
+            criteria = {self._key_attr: key}
+            attrlist = ['dn', self._key_attr]
+            res = self.context.search(criteria=criteria, attrlist=attrlist)
+            if not res:
+                raise KeyError(key)
+            if len(res) > 1:
+                msg = u'More than one principal with id "{0}" found.'
+                logger.warning(msg.format(key))
+            # XXX: as soon as LDAPNode.search result format has been changed,
```
Comment (rnixx) [inline src/node/ext/ldap/ugm/_api.py]: See related TODO note https://github.com/bluedynamics/node.ext.ldap/pull/12#discussion_r45972373

File: src/node/ext/ldap/ugm/_api.py [hunk 6]
```diff
@@ -505,10 +466,30 @@ def __getitem__(self, key):
...
+            #      this needs to be adopted
+            prdn = res[0][0]
+            if prdn in self.context._deleted_children:
+                raise KeyError(key)
+            dn = res[0][1]['dn']
+            # XXX: use explode_dn
```
Comment (rnixx) [inline src/node/ext/ldap/ugm/_api.py]: should go into 1.0, fine for beta

**Comment by jensens** [comment]
Idea: Prepare for beta status

File: src/node/ext/ldap/ugm/_api.py [hunk 7]
```diff
@@ -516,32 +497,39 @@ def __getitem__(self, key):
     @default
     @locktree
     def __iter__(self):
-        return self.context.__iter__()
+        attrlist = [self._key_attr]  # XXX: include RDN
+        res = self.context.search(attrlist=attrlist)
+        for principal in res:
+            # XXX: as soon as LDAPNode.search result format has been changed,
```
Comment (rnixx) [inline src/node/ext/ldap/ugm/_api.py]: See related TODO note

File: src/node/ext/ldap/ugm/_api.py [hunk 8]
```diff
@@ -629,10 +623,31 @@ def search(self, criteria=None, attrlist=None,
     @default
     @locktree
     def create(self, pid, **kw):
-        vessel = AttributedNode()
+        # XXX: mechanism for defining a target container if scope is SUBTREE
```
Comment (rnixx) [inline src/node/ext/ldap/ugm/_api.py]: Should this go into 1.0?

**Comment by jensens** [comment]
re-enabled README.rst test and got this
https://travis-ci.org/bluedynamics/node.ext.ldap/builds/93352109#L3292

**Pull request merged**